core/pta/attestation.c

```diff
@@ -55,21 +55,43 @@ static int cmp_regions(const void *a, const void *b)
 	return memcmp((void *)r1->va, (void *)r2->va, r1->size);
 }
 
-static TEE_Result hash_update(void *ctx, struct vm_region *r, uint32_t mode)
+#define TO_LE64(x) TEE_U64_BSWAP(TEE_U64_TO_BIG_ENDIAN(x))
+
+static TEE_Result hash_update_mode_tags(void *ctx, struct vm_region *r)
 {
-	struct file *f = NULL;
+	uint64_t offs_and_size[2] = { 0, TO_LE64(r->mobj->size) };
+	struct fobj *fobj = r->mobj->ops->get_fobj(r->mobj);
+	struct file *f = to_file_may_fail(r->mobj);
+	TEE_Result res = TEE_SUCCESS;
+	unsigned int poffs = 0;
 	uint8_t *tag = NULL;
 
+	/* Regions we're interested in should all have a struct file */
+	assert(f);
+	file_lock(f);
+	tag = file_get_tag(f);
+	assert(tag);
+	res = crypto_hash_update(ctx, tag, FILE_TAG_SIZE);
+	if (!res) {
+		file_unlock(f);
+		return res;
+	}
+	res = file_find_page_offset(f, fobj, &poffs);
+	file_unlock(f);
+	if (!res)
+		return res;
+	offs_and_size[0] = TO_LE64(poffs * SMALL_PAGE_SIZE);
```

Comment:
This looks quite ambitious. Preparing for a big endian OP-TEE with little endian TAs?

Reply:
Yeah I thought about a big endian TEE core indeed... Probably overkill, it probably doesn't even matter if the hash changes in this case... However I'll keep the `uint64_t` conversions so that a 32-bit TA can have the same hash whatever the size of `size_t` in the TEE (Aarch32 vs. Aarch64 core).

Reply:
Makes sense.